# RDF Metadata Profile Validation with Shape Expression:

## the Covid-19 sample metadata use case 

___

<div class="row">

  <div class="column">
    <div class="card">
      <div class="container">
        <i class="fa fa-qrcode fa-2x" style="color:#7e0038"></i>
        <h4><b>Recipe metadata</b></h4>
        <p> identifier: <a href="">RX.X</a> </p>
        <p> version: <a href="">v1.0</a> </p>
      </div>
    </div>
  </div>
  <div class="column">
    <div class="card">
      <div class="container">
        <i class="fa fa-fire fa-2x" style="color:#7e0038"></i>
        <h4><b>Difficulty level</b></h4>
        <i class="fa fa-fire fa-lg" style="color:#7e0038"></i>
        <i class="fa fa-fire fa-lg" style="color:#7e0038"></i>
        <i class="fa fa-fire fa-lg" style="color:#7e0038"></i>
        <i class="fa fa-fire fa-lg" style="color:lightgrey"></i>
        <i class="fa fa-fire fa-lg" style="color:lightgrey"></i>
  <!--       <p><span data-v-013baba1="" title="" class=""><svg data-v-013baba1="" viewBox="0 0 16 16" width="1em" height="1em" focusable="false" role="img" alt="icon" xmlns="http://www.w3.org/2000/svg" fill="currentColor" class="bi-bar-chart-fill b-icon bi medium-level"><g data-v-013baba1=""><rect width="4" height="5" x="1" y="10" rx="1"></rect><rect width="4" height="9" x="6" y="6" rx="1"></rect><rect width="4" height="14" x="11" y="1" rx="1"></rect></g></svg> Medium </span></p> -->
      </div>
    </div>
  </div>  
  <div class="column">
    <div class="card">
      <div class="container">
        <i class="fa fa-clock-o fa-2x" style="color:#7e0038;"></i>
        <h4><b>Reading Time</b></h4>
        <p><i class="fa fa-clock-o fa-lg" style="color:#7e0038;"></i> 30 minutes</p>
        <h4><b>Recipe Type</b></h4>
        <p><i class="fa fa-laptop fa-lg" style="color:#7e0038;"></i> Hands-on</p>
        <h4><b>Executable Code</b></h4>
        <p><i class="fa fa-play-circle fa-lg" style="color:#7e0038;"></i> Yes</p>
      </div>
    </div>
  </div>
  <div class="column">
    <div class="card">
      <div class="container">
        <i class="fa fa-group fa-2x" style="color:#7e0038;"></i>
        <h4><b>Intended Audience</b></h4>
        <p> <i class="fa fa-user-md fa-lg" style="color:#7e0038;"></i> Principal Investigators </p>
        <p> <i class="fa fa-database fa-lg" style="color:#7e0038;"></i> Data Managers </p>
        <p> <i class="fa fa-wrench fa-lg" style="color:#7e0038;"></i> Data Scientists </p>
<!--         <p> <i class="fa fa-money fa-lg" style="color:#7e0038;"></i> Funders</p> -->
      </div>
    </div>
  </div>
</div>

___


## Introduction:

>:information_source: 
This recipe is adapted from work carried during the Elixir Covid-19 biohackathon, by the ontology and workflow tracks, presented [here](http://covid19.genenetwork.org/) and detailed in the [following manuscript](https://github.com/arvados/bh20-seq-resource/blob/master/paper/paper.md) while all the code and associated material is hosted on this [github repository](https://github.com/arvados/bh20-seq-resource).


>:information_source: 
[Robert Hoendorf](https:// "title"), [Jose Emilio Labra Gayo](https:// "title"),[Thomas Liener](https:// "title"), [Nuria Queralt Rosinach](https:// "title")
, [Tazro  Ohta](https:// "title"), [Philippe Rocca-Serra](https:// "title"), [Claus Weilland](https:// "title"), [Piotr Prins](https:// "title"),  [Danielle Welter](https:// "title"). Thomas Liener  and Danielle Welter acted as coordinator between the ontology track and the workflow track led by Piotr Prins. 



In this specific report, we focus solely on the specific task of creating `covid-19 virus sample metadata reporting profile`. The aim is to ensure that each sequencing file generated by the sequencing efforts comes with sufficient descriptive metadata to allow basic correlation analysis. 
A `metadata reporting profile` should therefore be understood as the functional implementation of a "Minimum Information (MI) Guidelines". Minimal Information Guidelines are specifications listing what `must`, `should` or `may` be part of a report about some entity. These specifications are often informal, only available as unstructured text. For a list of know MI guidelines, consult the [FAIRsharing](https://fairsharing.org) catalogue, an Elixir-UK Resource. 


Therefore 3 essential steps were performed:
> * Listing essential sample attributes
> * Performing a semantic anchoring of these attributes
> * Defining a formal representation capturing those requirements
> * Expressing instance data in RDF/linked data format
> * Validating RDF instance data against requirements using a Shape Expression(SHEX)
> * Testing query cases by formulating SPARQL queries

The following sections detail each of these steps

## Defining the metadata fields

In the context of reporting viral genome sequence, the [Genome Standards Consortium (GSC)](https://gsc.org)'s Minimal Information about any Genome (MIXS) checklist is of relevance, and a record is available from FAIRsharing: [https://fairsharing.org/10.25504/FAIRsharing.9aa0zp](https://fairsharing.org/10.25504/FAIRsharing.9aa0zp).

<div><img src="https://i.imgur.com/8Y0U4Ai.png" width="750" border="1"/></div>

GSC's MIxS checklist identifies a number of key properties that should be reported when describing samples from which sequence data is generated. For each property, a unique `tag` has been agreed upon and vetted by the International Nucleotide Sequence Database Collaboration (INSDC). The INSDC is a long-standing foundational initiative that operates between DDBJ, EMBL-EBI and NCBI databases. INSDC covers the spectrum of data raw reads, through alignments and assemblies to functional annotation, enriched with contextual information relating to samples and experimental configurations. These tags are used to structure information in those genome archives. Harmonization is necessary as data exchange and mirroring is carried out between the 3 nodes.


Based on the GSC profile for a viral sample, the first step is to anchor those INSDC tags to one or more semantic frameworks.

<div><img src="https://i.imgur.com/lwV1cPd.png" width="300" border="1"/></div>
<div><img src="https://i.imgur.com/1ahQLjy.png" width="300" border="1"/></div> 


The use case and challenge at hand in the context of the Covid-19 Biohackathon was to develop a procedure which would allow, on the on hand, the capture of viral sequence information, 
and on the other hand, the simultaneous harvesting of critical sample descriptive metadata but in both cases, this should be done RDF/linked data, with information persistence to a SPARQL endpoint.

This therefore presents a number of tasks to complete in order to realize that objective. The following sections detail each of the key steps.



## Semantic anchoring of metadata element to RDF types / Ontology Classes:

> From the set of INSDC tags covering in part the ENA annotation requirements and the MIxS requirements, several mappings into different semantic frameworks have been made by the developers.
These semantic frameworks are resources from the [OBO Foundry](https://obofoundry.org), [Wikidata](https://wikidata.org) or [Schema.org](https://schema.org).
> For the final implementation, only the OBO related mappings has been used.
> The principle remains the same for creating a validation profile.

<!-- ![](https://i.imgur.com/Ro92a7D.png) -->

<div><img src="https://i.imgur.com/Ro92a7D.png" width="800" border="1"/></div>


For the rest of the development, the OBO based mapping were retained.

<div><img src="https://i.imgur.com/Lc7FcPs.png" width="800" border="1"/></div>

<!-- ![](https://i.imgur.com/Lc7FcPs.png) -->

From then on, the formalization process can test. The syntax chose is YAML in this example.


## 1. definition of a metadata acquisition form with YAML:

YAML stands for `YAML Ain't Markup Language` and defines a `human-readable data-serialization language`. This very simple specification allows to define `keys` and associated typed values to them.
The keys can be viewed as the main properties in a data structure blueprint. The values associated with a key correspond to instance level information of those object declarations. Without going into too much details

The annotation requirement has been captured in a YAML document, which is also used to build a web form (see below). The full document is available from:

Below is a partial view of the metadata yaml form, showing how `host` information requirements have been defined:


```bash
$base: http://biohackathon.org/bh20-seq-schema
$namespaces:
  sch: https://schema.org/
  efo: http://www.ebi.ac.uk/efo/
  obo: http://purl.obolibrary.org/obo/
  sio: http://semanticscience.org/resource/
  edam: http://edamontology.org/
  evs: http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#

$graph:

- name: hostSchema
  type: record
  fields:
    host_species:
        doc: Host species as defined in NCBITaxon, e.g. http://purl.obolibrary.org/obo/NCBITaxon_9606 for Homo sapiens
        type: string
        jsonldPredicate:
          _id: http://www.ebi.ac.uk/efo/EFO_0000532
          _type: "@id"
          noLinkCheck: true
    host_id:
        doc: Identifer for the host. If you submit multiple samples from the same host, use the same host_id for those samples
        type: string?
        jsonldPredicate:
          _id: http://semanticscience.org/resource/SIO_000115
    host_sex:
        doc: Sex of the host as defined in PATO, expect Male (http://purl.obolibrary.org/obo/PATO_0000384) or Female (http://purl.obolibrary.org/obo/PATO_0000383) or in Intersex (http://purl.obolibrary.org/obo/PATO_0001340)
        type: string?
        jsonldPredicate:
          _id: http://purl.obolibrary.org/obo/PATO_0000047
          _type: "@id"
          noLinkCheck: true
    host_age:
        doc: Age of the host as number (e.g. 50)
        type: int?
        jsonldPredicate:
          _id: http://purl.obolibrary.org/obo/PATO_0000011
    host_age_unit:
        doc: Unit of host age e.g. http://purl.obolibrary.org/obo/UO_0000036
        type: string?
        jsonldPredicate:
          _id: http://purl.obolibrary.org/obo/NCIT_C42574
          _type: "@id"
          noLinkCheck: true
    host_health_status:
        doc: A condition or state at a particular time, must be one of the following (obo:NCIT_C115935 obo:NCIT_C3833 obo:NCIT_C25269 obo:GENEPIO_0002020 obo:GENEPIO_0001849 obo:NCIT_C28554 obo:NCIT_C37987)
        type: string?
        jsonldPredicate:
          _id: http://purl.obolibrary.org/obo/NCIT_C25688
          _type: "@id"
          noLinkCheck: true
    host_treatment:
      doc: Process in which the act is intended to modify or alter host status
      type: string?
      jsonldPredicate:
          _id: http://www.ebi.ac.uk/efo/EFO_0000727
```

source: https://github.com/arvados/bh20-seq-resource/blob/master/bh20sequploader/bh20seq-schema.yml


### The corresponding metadata acquisition web form:

<div><img src="https://i.imgur.com/5eQN9hw.png" width="800" border="1"/></div>

<!-- ![](https://i.imgur.com/5eQN9hw.png) -->

<div><img src="https://i.imgur.com/RR5GSgi.png" width="800" border="1"/></div>

<!-- ![](https://i.imgur.com/RR5GSgi.png) -->



## 2. Exemplar instance data:


When users submit information via the form (or by other programatic means), a instance yaml file is generated, which looks like this:


```bash
id: placeholder

host:
    host_id: XX1
    host_species: http://purl.obolibrary.org/obo/NCBITaxon_9606
    host_sex: http://purl.obolibrary.org/obo/PATO_0000384
    host_age: 20
    host_age_unit: http://purl.obolibrary.org/obo/UO_0000036
    host_health_status: http://purl.obolibrary.org/obo/NCIT_C25269
    host_treatment: Process in which the act is intended to modify or alter host status (Compounds)
    host_vaccination: [vaccines1,vaccine2]
    ethnicity: http://purl.obolibrary.org/obo/HANCESTRO_0010
    additional_host_information: Optional free text field for additional information

sample:
    sample_id: Id of the sample as defined by the submitter 
    collector_name: Name of the person that took the sample
    collecting_institution: Institute that was responsible of sampling  
    specimen_source: [http://purl.obolibrary.org/obo/NCIT_C155831,http://purl.obolibrary.org/obo/NCIT_C155835]
    collection_date: "2020-01-01"
    collection_location: http://www.wikidata.org/entity/Q148
    sample_storage_conditions: frozen specimen
    source_database_accession: [http://identifiers.org/insdc/LC522350.1#sequence]
    additional_collection_information: Optional free text field for additional information

virus:
    virus_species: http://purl.obolibrary.org/obo/NCBITaxon_2697049
    virus_strain: SARS-CoV-2/human/CHN/HS_8/2020

technology:
    sample_sequencing_technology: [http://www.ebi.ac.uk/efo/EFO_0009173,http://www.ebi.ac.uk/efo/EFO_0009173]
    sequence_assembly_method: Protocol used for assembly
    sequencing_coverage: [70.0, 100.0]
    additional_technology_information: Optional free text field for additional information

submitter:
    authors: [John Doe, Joe Boe, Jonny Oe]
    submitter_name: [John Doe]
    submitter_address: John Doe\'s address
    originating_lab: John Doe kitchen
    lab_address: John Doe\'s address
    provider_sample_id: XXX1
    submitter_sample_id: XXX2
    publication: PMID00001113
    submitter_orcid: [https://orcid.org/0000-0000-0000-0000,https://orcid.org/0000-0000-0000-0001]
    additional_submitter_information: Optional free text field for additional information

```

source: https://github.com/arvados/bh20-seq-resource/blob/master/example/maximum_metadata_example.yaml



## 3. Conversion to RDF and entry validation based on SHEX expression :

In the previous steps, we use YAML to define an annotation schema and instance data, which essentially associated values to the schema keys are also stored as a YAML document.
The `magic sauce` to transform these YAML instance documents to RDF is a python component developed in the context of the CWL language and Apache SALAD schema:https://www.commonwl.org/v1.0/SchemaSalad.html



Working with a domain expert ([Dr Jose Emilio Labra Gayo - (Oviedo Uni)](http://di002.edv.uniovi.es/~labra/), using [Shape Expression syntax](https://shex.io/shex-semantics/) the following profile was developed and used to validate the RDF before persistence to the SPARQL endpoint:


```bash
PREFIX : <https://raw.githubusercontent.com/arvados/bh20-seq-resource/master/bh20sequploader/bh20seq-shex.rdf#>
PREFIX MainSchema: <http://biohackathon.org/bh20-seq-schema#MainSchema/>
PREFIX hostSchema: <http://biohackathon.org/bh20-seq-schema#hostSchema/>
PREFIX xsd: <http://www.w3.org/2001/XMLSchema#>
PREFIX obo: <http://purl.obolibrary.org/obo/>
PREFIX sio: <http://semanticscience.org/resource/>
PREFIX efo: <http://www.ebi.ac.uk/efo/>
PREFIX evs: <http://ncicb.nci.nih.gov/xml/owl/EVS/Thesaurus.owl#>
PREFIX edam: <http://edamontology.org/>
PREFIX wikidata: <http://www.wikidata.org/entity/>

:submissionShape {
  MainSchema:host   @:hostShape ;
  MainSchema:sample @:sampleShape ;
  MainSchema:submitter @:submitterShape ;
  MainSchema:technology @:technologyShape ;
  MainSchema:virus @:virusShape;
}

:hostShape  {
    efo:EFO_0000532 [ obo:NCBITaxon_~ ] ;
    sio:SIO_000115 xsd:string ?;
    obo:PATO_0000047 [ obo:PATO_0000384 obo:PATO_0000383 obo:PATO_0001340] ?;
    obo:PATO_0000011 xsd:integer ?;
    obo:NCIT_C42574 [ obo:UO_~ ] ?;
  obo:NCIT_C25688 [obo:NCIT_C115935 obo:NCIT_C3833 obo:NCIT_C25269 obo:GENEPIO_0002020 obo:GENEPIO_0001849 obo:NCIT_C28554 obo:NCIT_C37987 ] ? ;
    efo:EFO_0000727 xsd:string ?;
    obo:VO_0000002 xsd:string {0,10};
    sio:SIO_001167 xsd:string ?;
    sio:SIO_001014 [ obo:HANCESTRO_~ ] ? ; #ethnicity
}

:sampleShape  {
    sio:SIO_000115 xsd:string;
    evs:C25164 xsd:string;
    obo:GAZ_00000448 [wikidata:~] ;
    obo:OBI_0001895 xsd:string ?;
    obo:NCIT_C41206 xsd:string ?;
    obo:OBI_0001479 IRI {0,2};
    obo:OBI_0001472 xsd:string ?;
    sio:SIO_001167 xsd:string ?;
  edam:data_2091 IRI {0,3};
}

:submitterShape {
    obo:NCIT_C42781 xsd:string + ;
    sio:SIO_000116 xsd:string *;
    sio:SIO_000172 xsd:string ?;
    obo:NCIT_C37984 xsd:string ?;
    obo:OBI_0600047 xsd:string ?;
    obo:NCIT_C37900 xsd:string ?;
    efo:EFO_0001741 xsd:string ?;
    obo:NCIT_C19026 xsd:string ?;
    sio:SIO_000115 
    sio:SIO_001167 xsd:string ?;
}

:technologyShape {
    obo:OBI_0600047 IRI {0,3} ;
    efo:EFO_0002699 xsd:string ?;
    obo:FLU_0000848 xsd:double OR xsd:integer {0,3};
    sio:SIO_001167 xsd:string ?;
}

:virusShape{
  edam:data_1875 [ obo:NCBITaxon_~ ] ;
    sio:SIO_010055 xsd:string ?;
}
```

source: https://github.com/arvados/bh20-seq-resource/blob/master/bh20sequploader/bh20seq-shex.rdf

Using the [WESO](http://www.weso.es/) developed [dedicated tool](http://rdfshape.weso.es/),  Shape Expression can be viewed. 


<div><img src="" link="http://rdfshape.weso.es/shExVisualize?activeSchemaTab=%23schemaTextArea&schema=PREFIX%20%3A%20%3Chttps%3A%2F%2Fraw.githubusercontent.com%2Farvados%2Fbh20-seq-resource%2Fmaster%2Fbh20sequploader%2Fbh20seq-shex.rdf%23%3E%0APREFIX%20MainSchema%3A%20%3Chttp%3A%2F%2Fbiohackathon.org%2Fbh20-seq-schema%23MainSchema%2F%3E%0APREFIX%20hostSchema%3A%20%3Chttp%3A%2F%2Fbiohackathon.org%2Fbh20-seq-schema%23hostSchema%2F%3E%0APREFIX%20xsd%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%3E%0APREFIX%20obo%3A%20%3Chttp%3A%2F%2Fpurl.obolibrary.org%2Fobo%2F%3E%0APREFIX%20sio%3A%20%3Chttp%3A%2F%2Fsemanticscience.org%2Fresource%2F%3E%0APREFIX%20efo%3A%20%3Chttp%3A%2F%2Fwww.ebi.ac.uk%2Fefo%2F%3E%0APREFIX%20evs%3A%20%3Chttp%3A%2F%2Fncicb.nci.nih.gov%2Fxml%2Fowl%2FEVS%2FThesaurus.owl%23%3E%0APREFIX%20edam%3A%20%3Chttp%3A%2F%2Fedamontology.org%2F%3E%0APREFIX%20wikidata%3A%20%3Chttp%3A%2F%2Fwww.wikidata.org%2Fentity%2F%3E%0A%0A%3AsubmissionShape%20%7B%0A%20%20MainSchema%3Ahost%20%20%20%40%3AhostShape%20%3B%0A%20%20MainSchema%3Asample%20%40%3AsampleShape%20%3B%0A%20%20MainSchema%3Asubmitter%20%40%3AsubmitterShape%20%3B%0A%20%20MainSchema%3Atechnology%20%40%3AtechnologyShape%20%3B%0A%20%20MainSchema%3Avirus%20%40%3AvirusShape%3B%0A%7D%0A%0A%3AhostShape%20%20%7B%0A%20%20%09efo%3AEFO_0000532%20%5B%20obo%3ANCBITaxon_~%20%5D%20%3B%0A%20%20%20%20sio%3ASIO_000115%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3APATO_0000047%20%5B%20obo%3APATO_0000384%20obo%3APATO_0000383%20%5D%20%3F%3B%0A%20%20%20%20obo%3APATO_0000011%20xsd%3Ainteger%20%3F%3B%0A%20%20%20%20obo%3ANCIT_C42574%20%5B%20obo%3AUO_~%20%5D%20%3F%3B%0A%09obo%3ANCIT_C25688%20%5Bobo%3ANCIT_C115935%20obo%3ANCIT_C3833%20obo%3ANCIT_C25269%20obo%3AGENEPIO_0002020%20obo%3AGENEPIO_0001849%20obo%3ANCIT_C28554%20obo%3ANCIT_C37987%20%5D%20%3F%20%3B%0A%20%20%20%20efo%3AEFO_0000727%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3AVO_0000002%20xsd%3Astring%20%7B0%2C10%7D%3B%0A%20%20%20%20sio%3ASIO_001167%20xsd%3Astring%20%3F%3B%0A%7D%0A%0A%3AsampleShape%20%20%7B%0A%20%20%20%20sio%3ASIO_000115%20xsd%3Astring%3B%0A%09evs%3AC25164%20xsd%3Astring%3B%0A%09obo%3AGAZ_00000448%20%5Bwikidata%3A~%5D%20%3B%0A%20%20%20%20obo%3AOBI_0001895%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3ANCIT_C41206%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3AOBI_0001479%20IRI%20%7B0%2C2%7D%3B%0A%20%20%20%20obo%3AOBI_0001472%20xsd%3Astring%20%3F%3B%0A%20%20%20%20sio%3ASIO_001167%20xsd%3Astring%20%3F%3B%0A%09edam%3Adata_2091%20IRI%20%7B0%2C3%7D%3B%0A%7D%0A%0A%3AsubmitterShape%20%7B%0A%20%20%20%20obo%3ANCIT_C42781%20xsd%3Astring%20%2A%20%3B%0A%20%20%20%20sio%3ASIO_000116%20xsd%3Astring%20%2A%3B%0A%20%20%20%20sio%3ASIO_000172%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3ANCIT_C37984%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3AOBI_0600047%20xsd%3Astring%20%3F%3B%0A%20%20%09obo%3ANCIT_C37900%20xsd%3Astring%20%3F%3B%0A%20%20%20%20efo%3AEFO_0001741%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3ANCIT_C19026%20xsd%3Astring%20%3F%3B%0A%20%20%20%20sio%3ASIO_000115%20%2Fhttps%3A%5Cu002F%5Cu002Forcid.org%5Cu002F.%7B4%7D-.%7B4%7D-.%7B4%7D-.%7B4%7D%2F%20%7B0%2C10%7D%3B%0A%20%20%20%20sio%3ASIO_001167%20xsd%3Astring%20%3F%3B%0A%7D%0A%0A%3AtechnologyShape%20%7B%0A%20%20%20%20obo%3AOBI_0600047%20IRI%20%7B0%2C3%7D%20%3B%0A%20%20%20%20efo%3AEFO_0002699%20xsd%3Astring%20%3F%3B%0A%20%20%20%20obo%3AFLU_0000848%20xsd%3Ainteger%20%7B0%2C2%7D%3B%0A%20%20%20%20sio%3ASIO_001167%20xsd%3Astring%20%3F%3B%0A%7D%0A%0A%3AvirusShape%7B%0A%09edam%3Adata_1875%20%5B%20obo%3ANCBITaxon_~%20%5D%20%3B%0A%20%20%09sio%3ASIO_010055%20xsd%3Astring%20%3F%3B%0A%7D&schemaEmbedded=false&schemaEngine=ShEx&schemaFormat=ShExC&schemaFormatTextArea=ShExC" width="800" border="1"></div>






In the example below a [schema.org](https://schema.org) base shex expression in presented.

<div><img src="https://i.imgur.com/z2rriQu.png" link="http://rdfshape.weso.es/shExVisualize?activeSchemaTab=%23schemaTextArea&schema=PREFIX%20%3A%20%3Chttp%3A%2F%2Fwhatever%2F%3E%0APREFIX%20MainSchema%3A%20%3Chttp%3A%2F%2Fbiohackathon.org%2Fbh20-seq-schema%23MainSchema%2F%3E%0APREFIX%20hostSchema%3A%20%3Chttp%3A%2F%2Fbiohackathon.org%2Fbh20-seq-schema%23hostSchema%2F%3E%0APREFIX%20xsd%3A%20%20%3Chttp%3A%2F%2Fwww.w3.org%2F2001%2FXMLSchema%23%3E%0APREFIX%20efo%3A%20%20%20%3Chttp%3A%2F%2Fwww.ebi.ac.uk%2Fefo%2F%3E%0Aprefix%20my%3A%20%3Chttp%3A%2F%2Fexample.org%2F%3E%0Aprefix%20ex%3A%20%3Chttp%3A%2F%2Fexample.org%2F%3E%0Aprefix%20rdf%3A%20%3Chttp%3A%2F%2Fwww.w3.org%2F1999%2F02%2F22-rdf-syntax-ns%23%3E%0Aprefix%20sdo%3A%20%3Chttp%3A%2F%2Fschema.org%2F%3E%0Aprefix%20geo%3A%20%3Chttp%3A%2F%2Fwww.opengis.net%2Font%2Fgeosparql%23%3E%0Aprefix%20sdo%3A%20%3Chttp%3A%2F%2Fschema.org%2F%3E%0A%0A%3ASubmissionShape%20%7B%0A%20%20MainSchema%3Ahost%20%20%20%20%20%20%20%40%3AHost%20%3B%20%0A%20%20MainSchema%3Asubmitter%20%20%40%3ASubmitter%20%3B%0A%20%20MainSchema%3Asample%20%20%20%20%20%40%3ASample%20%3B%0A%20%20MainSchema%3Asubmitter%20%20%40%3ASubmitter%20%3B%0A%20%20MainSchema%3Atechnology%20%40%3ATechnology%20%3B%0A%7D%0A%20%20%0A%3ASubmitter%20%7B%0A%20%20%09sdo%3AinfectiousAgentClass%20xsd%3Astring%3B%0A%20%20%09sdo%3AinfectiousAgent%20xsd%3Astring%0A%7D%20%0A%20%20%0A%3AHost%20%7B%0A%20a%20%5B%20sdo%3APerson%20%20%0A%20%20%20%20%20sdo%3APatient%20%0A%20%20%20%20%20sdo%3AAnimal%20%20%20%23not%20defined%20yet%20in%20sdo%0A%20%5D%3B%20%0A%20sdo%3Aidentifier%20IRI%2B%3B%0A%20sdo%3Aname%20xsd%3Astring%3B%0A%20sdo%3Agender%20%5B%22female%22%20%22male%22%20%22other%22%20%22NA%22%5D%3B%0A%20sdo%3Aage%20xsd%3Ainteger%3B%0A%20sdo%3Aunit%20IRI%2B%3B%0A%20sdo%3AhealthCondition%20IRI%2B%3B%0A%20sdo%3Adrug%20%40%3ADrug%20%2B%3B%0A%20sdo%3AprimaryPrevention%20%40%3APrimaryPrevention%20%3B%0A%20sdo%3Acomment%20xsd%3Astring%20%2B%0A%7D%20%20%0A%20%0A%3ADrug%20IRI%0A%0A%3APrimaryPrevention%20%7B%0A%7D%0A%0A%3ASample%20%7B%0A%20%20%09sdo%3AdateCreated%20xsd%3ADate%3B%0A%20%20%09sdo%3AfromLocation%20IRI%2B%3B%0A%20%20%09sdo%3AfromLocation%20%40geo%3AGeoLocation%3B%20%23to%20defined%20or%20fetch%20from%20existing%0A%20%20%09sdo%3Aidentifier%20IRI%3B%0A%20%20%09sdo%3Asender%20IRI%0A%7D%20%0A%0A%3ASubmitter%20%7B%0A%20%20%09rdf%3Atype%20%5B%20sdo%3APerson%20%5D%3B%0A%20%20%09sdo%3Aidentifier%20IRI%3B%0A%20%20%09sdo%3Afullname%20xsd%3Astring%3B%0A%20%20%09sdo%3Arolename%20xsd%3Astring%3B%0A%20%20%7D%20%0A%0A%3ATechnology%20%7B%0A%20%20%09sdo%3AmeasurementTechnique%20IRI%2B%20%3B%0A%20%20%09sdo%3AmedicalDevice%20IRI%2B%3B%0A%7D%20%0A%0Ageo%3AGeoLocation%20%7B%7D&schemaEmbedded=false&schemaEngine=ShEx&schemaFormat=ShExC&schemaFormatTextArea=ShExC" width="800" border="1"></div>

<!-- ![](https://i.imgur.com/z2rriQu.png) -->

There is a blog focusing mainly on the sequence analysis but there is a section on metadata validation.

## 4. SPARQL queries available here:

http://covid19.genenetwork.org/blog?id=using-covid-19-pubseq-part1

### 4.1. The SPARQL endpoint

The following endpoint during the Elixir Covid-19 Biohackthon and metadata information converted from the YAML definition to RDF turtle format was loaded in the following SPARQL Endpoint.

http://sparql.genenetwork.org/sparql/

The collection of metadata in rdf format is available for download: https://collections.lugli.arvadosapi.com/c=lugli-4zz18-z513nlpqm03hpca/mergedmetadata.ttl

### 4.2. Exploring the metadata described the FASTQ sequence files

Limiting search to metadata add http://covid-19.genenetwork.org/graph/metadata.ttl in the top input box. Now you can find a predicate for submitter that looks like http://biohackathon.org/bh20-seq-schema#MainSchema/submitter.


```bash
PREFIX pubseq: <http://biohackathon.org/bh20-seq-schema#MainSchema/>
PREFIX sio: <http://semanticscience.org/resource/>
select distinct ?sample ?p ?o
{
   ?sample sio:SIO_000115 "MT326090.1" .
   ?sample ?p ?o .
}
```


---

## Conclusions:

> In this recipe, we have presented how to implement a minimal medata profile and validate data entry with a specific technology stack: namely using RDF and Shape Express standard. Other approaches are possible and we provide details in a dedicated recipe where JSON schema and JSON-LD technologies are used. 
> This recipe tackles an important aspect of the FAIR principles, shining the light on the need to provide sufficient descriptive metadata to associate with an assay data file to allow its correct interpretation.
> The recipe therefore provides a piece of the jigsaw to establish a FAIR datasets.
> There are some caveats or improvements which could be made. For instance, the devised shex expression and the associated instance RDF graph  could be assigned a persistent identifiers (PID).
> Another improvement could be a better integration with repositories such FAIRsharing or the main submission systems to
>
> #### What should I read next?
> * [How to validate a FASTQ sequencing file?]()
> * [How to express Minimal Metadata checklist in machine readable format]()
> * [How to validate metadata with JSON Schema?]()
> * [How to perform data integration with SPARQL?]()


## FAIRification Objectives, Inputs and Outputs

| Actions.Objectives.Tasks  | Input | Output  |
| :------------- | :------------- | :------------- |
| [validation]()  | [text]()  | [DOI]()  |
||file||

## Table of Data Standards

| Data Formats  | Terminologies | Models  |
| :------------- | :------------- | :------------- |
| [YAML]()  | [EFO]()  |   |
| [RDF]()|[SIO]()||
| [SPARQL]()|[SDO](https://schema.org)||
| |[EDAM]()||
| |[OBO]()||

___ 

## Biobliographic reference:

{% bibliography --cited %}

___
## Authors:

| Name | Affiliation  | orcid | CrediT role  |
| :------------- | :------------- | :------------- |:------------- |
| Philippe Rocca-Serra |  University of Oxford, Data Readiness Group| [0000-0001-9853-5668](https://orcid.org/orcid.org/0000-0001-9853-5668) | Writing - Original Draft, Shex expression, ontology mapping |
| Danielle Welter |  University of Luxembourg | xxxx-xxxx-xxxx-xxx| Review, ontology mapping| 
| Jose Emilio Labra Gayo |  University of Oviedo | xxxx-xxxx-xxxx-xxx| Shex Expression | 
|Robert Giessmann|Bayer AG||Review|
||||


___
## License:

<a href="https://creativecommons.org/licenses/by/4.0/"><img src="https://mirrors.creativecommons.org/presskit/buttons/80x15/png/by-sa.png" height="20"/></a>
